android/src/main/java/com/twiliorn/library/Utils.java

```diff
@@ -0,0 +1,131 @@
+
+import tvi.webrtc.VideoFrame;
+
+public final class Utils {
+    private Utils() {}
+
+    /**
+     * TODO: refactor to use more optimized Streem code: https://github.com/streem/streem-app/blob/main/android/StreemSDK/src/main/java/pro/streem/sdk/internal/util/ImageUtil.kt
+     * Create an NV21Buffer with the same pixel content as the given I420 buffer. Ripped from here: https://stackoverflow.com/a/55158228
+     */
+    private static byte[] createNV21Data(VideoFrame.I420Buffer i420Buffer) {
+        final int width = i420Buffer.getWidth();
+        final int height = i420Buffer.getHeight();
+        final int chromaStride = width;
+        final int chromaWidth = (width + 1) / 2;
+        final int chromaHeight = (height + 1) / 2;
+        final int ySize = width * height;
+        final ByteBuffer nv21Buffer = ByteBuffer.allocateDirect(ySize + chromaStride * chromaHeight);
+        // We don't care what the array offset is since we only want an array that is direct.
+        @SuppressWarnings("ByteBufferBackingArray") final byte[] nv21Data = nv21Buffer.array();
+        for (int y = 0; y < height; ++y) {
+            for (int x = 0; x < width; ++x) {
+                final byte yValue = i420Buffer.getDataY().get(y * i420Buffer.getStrideY() + x);
+                nv21Data[y * width + x] = yValue;
+            }
+        }
+        for (int y = 0; y < chromaHeight; ++y) {
+            for (int x = 0; x < chromaWidth; ++x) {
+                final byte uValue = i420Buffer.getDataU().get(y * i420Buffer.getStrideU() + x);
+                final byte vValue = i420Buffer.getDataV().get(y * i420Buffer.getStrideV() + x);
+                nv21Data[ySize + y * chromaStride + 2 * x + 0] = vValue;
+                nv21Data[ySize + y * chromaStride + 2 * x + 1] = uValue;
+            }
+        }
+        return nv21Data;
+    }
+
+    // i420 -> nv21 -> yuv -> jpeg
+    public static void saveVideoFrame(VideoFrame frame, Context context) {
+        Log.d(TwilioPackage.TAG, "saving video frame");
+        String fileId = UUID.randomUUID().toString();
+        String filename = "rntframe-" + fileId + ".jpeg";
+        Log.d(TwilioPackage.TAG, "saving frame for file " + filename);
+
+        VideoFrame.I420Buffer i420Buffer = frame.getBuffer().toI420();
+        final int width = i420Buffer.getWidth();
+        final int height = i420Buffer.getHeight();
+        Log.w(TwilioPackage.TAG, "frame width: " + width + ", height: " + height);
+
+        byte[] nv21Data = createNV21Data(i420Buffer);
+
+        YuvImage yuvImage = new YuvImage(nv21Data, ImageFormat.NV21, width, height, null);
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        yuvImage.compressToJpeg(new Rect(0, 0, width, height), 90, out);
+        byte[] rawImageBytes = out.toByteArray();
+        byte[] imageBytes = rotateJPEGByteArray(rawImageBytes, frame.getRotation());
+
+        try (FileOutputStream fos = context.openFileOutput(filename, Context.MODE_PRIVATE)) {
```

Comment:
Seeing now how you're using `context`, I think for `openFileOutput()` you only need the top-level application context. So to avoid the context leak I mentioned in my other comment, you could pass in `context.getApplicationContext()` into `saveVideoFrame()`. The application context is guaranteed to be around as long as the app itself is running, so there's no danger of leaking it.

Reply:
Ok that makes sense to me.